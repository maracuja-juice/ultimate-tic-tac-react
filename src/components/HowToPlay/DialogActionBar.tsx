import { DialogButton } from '@rmwc/dialog';
import { ArrowButtons } from '../ArrowButtons/ArrowButtons';
import { Button } from '@rmwc/button';
import { ArrowRightIcon } from '../Icons';
import * as React from 'react';

interface DialogActionBarProps {
  onClose: () => void;
  onForward: () => void;
  onBackward: () => void;
  stepNumber: number;
  maxStepNumber: number;
}

export function DialogActionBar(props: DialogActionBarProps) {
  const { onClose, onForward, onBackward, stepNumber, maxStepNumber } = props;
  const isLastStep = stepNumber === maxStepNumber;
  const isFirstStep = stepNumber === 0;

  return (
    <>
      {isFirstStep && <DialogButton action="close">Cancel</DialogButton>}
      <ArrowButtons
        value={stepNumber}
        maxValue={maxStepNumber}
        minValue={0}
        onInteraction={(forward) => {
          forward ? onForward() : onBackward();
        }}
        handleKeyboard={true}
        leftButtonConfig={{ buttonProps: { raised: false }, hide: isFirstStep }}
        rightButtonConfig={{
          buttonProps: { raised: false },
          hide: isLastStep,
        }}
      />
      {isLastStep && (
        <Button
          raised={true}
          trailingIcon={{ icon: <ArrowRightIcon />, 'aria-hidden': true }}
          onClick={onClose}
        >
          Let's Play
        </Button>
      )}
    </>
  );
}
