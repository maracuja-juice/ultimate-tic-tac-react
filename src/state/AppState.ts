import { Point } from '../util';

export interface AppState {
  currentGame: GameInformation;
  finishedGames: FinishedGameState[];
  analysisGame: AnalysisGame;
}

export interface GameInformation {
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
}

export interface FinishedGameState {
  id?: string;
  date: string;
  winner: Winner;
  gameState: SmallBoardInformation[];
  moves: MoveState[];
  saveState: string;
  errorMessage: string;
}

export interface AnalysisGame {
  id: string;
  board: SmallBoardInformation[];
  game: GameState;
  moves: MoveState[];
  activeBoards: Point[];
  currentMove: number;
}

export interface GameState {
  currentPlayer: Player;
  isFinished: boolean;
  winningPlayer: Winner;
}

export interface TileInformation {
  position: Point;
  value: TileValue;
}

export interface SmallBoardInformation extends TileInformation {
  tiles: SmallTileInformation[];
}

export interface SmallTileInformation extends TileInformation {
  boardPosition: Point;
}

export enum TileValue {
  Cross = 0,
  Circle = 1,
  Empty = 2,
  Destroyed = 3,
}

export enum Winner {
  Cross = 0,
  Circle = 1,
  None = 2,
  Draw = 3,
}

export enum Player {
  Cross = 0,
  Circle = 1,
}

export interface Move {
  boardPosition: Point;
  tilePosition: Point;
}

export interface MoveState extends Move {
  player: Player;
  moveNumber: number;
}

export interface GenericAction {
  type: string;
  // tslint:disable-next-line: no-any
  payload?: any;
}

export interface MarkSpecially {
  condition: boolean;
  position?: { boardPosition: Point; tilePosition: Point };
}
