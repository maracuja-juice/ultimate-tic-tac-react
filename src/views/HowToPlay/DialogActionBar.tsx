import { DialogButton } from '@rmwc/dialog';
import { Button } from '@rmwc/button';
import * as React from 'react';
import { ArrowButtons } from '../../components/ArrowButtons/ArrowButtons';
import { ArrowRightIcon } from '../../components/Icons';

interface DialogActionBarProps {
  onClose: () => void;
  onForward: () => void;
  onBackward: () => void;
  stepNumber: number;
  maxStepNumber: number;
}

export function DialogActionBar(props: DialogActionBarProps) {
  const { onClose, onForward, onBackward, stepNumber, maxStepNumber } = props;
  const isLastStep = stepNumber === maxStepNumber;
  const isFirstStep = stepNumber === 0;

  return (
    <>
      {isFirstStep && <DialogButton action="close">Cancel</DialogButton>}
      <ArrowButtons
        value={stepNumber}
        maxValue={maxStepNumber}
        minValue={0}
        onInteraction={(forward) => {
          forward ? onForward() : onBackward();
        }}
        handleKeyboard={true}
        leftButtonConfig={{ buttonProps: { raised: false }, hide: isFirstStep }}
        rightButtonConfig={{
          buttonProps: { raised: false },
          hide: isLastStep,
        }}
      />
      {isLastStep && (
        <Button
          raised={true}
          trailingIcon={{ icon: <ArrowRightIcon />, 'aria-hidden': true }}
          onClick={onClose}
        >
          Let's Play
        </Button>
      )}
    </>
  );
}
