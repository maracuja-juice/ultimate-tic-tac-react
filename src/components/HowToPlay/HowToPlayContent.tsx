import * as React from 'react';
import { HowToPlayBoardState } from '../../state/AppState';
import { BigBoard } from '../Board/BigBoard/BigBoard';

interface HowToPlayContentProps {
  boardState: HowToPlayBoardState;
}

export function HowToPlayContent(props: HowToPlayContentProps) {
  const { boardState } = props;
  return (
    <BigBoard
      animate={boardState.animate}
      currentPlayer={boardState.currentPlayer}
      board={boardState.board}
      activeBoards={boardState.activeBoards}
    />
  );
}
