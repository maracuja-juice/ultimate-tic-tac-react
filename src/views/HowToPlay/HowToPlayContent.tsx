import * as React from 'react';
import { HowToPlayBoardState } from '../../state/AppState';
import { BigBoard } from '../../components/Board/BigBoard/BigBoard';

interface HowToPlayContentProps {
  boardState: HowToPlayBoardState;
}

export function HowToPlayContent({ boardState }: HowToPlayContentProps) {
  return (
    <BigBoard
      animate={boardState.animate}
      currentPlayer={boardState.currentPlayer}
      board={boardState.board}
      activeBoards={boardState.activeBoards}
    />
  );
}
