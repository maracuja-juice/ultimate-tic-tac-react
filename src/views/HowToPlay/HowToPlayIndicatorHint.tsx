import * as React from 'react';
import styles from './HowToPlayDialog.module.css';
import { Tile } from '../../components/Board/Tile/Tile';
import { TileValue } from '../../state/AppState';

export function HowToPlayIndicatorHint() {
  return (
    <div className={styles.howToPlayIndicator}>
      <Tile
        value={TileValue.Empty}
        isTileRound={false}
        clickable={true}
        animate={false}
      />
    </div>
  );
}
