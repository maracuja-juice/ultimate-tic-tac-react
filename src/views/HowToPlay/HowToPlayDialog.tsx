import * as React from 'react';
import {
  Dialog,
  DialogActions,
  DialogContent,
  DialogTitle,
} from '@rmwc/dialog';
import styles from './HowToPlayDialog.module.css';
import { HowToPlayContent } from './HowToPlayContent';
import { HowToPlayBoardState } from '../../state/AppState';
import { DialogActionBar } from './DialogActionBar';
import { useKeyPressEvent } from 'react-use';
import { Keys } from '../../lib';

export interface HowToPlayDialogProps {
  onClose: () => void;
  onForward: () => void;
  onBackward: () => void;
  stepNumber: number;
  maxStepNumber: number;
  text: React.ReactNode;
  boardState: HowToPlayBoardState;
}

export function HowToPlayDialog(props: HowToPlayDialogProps) {
  const {
    onClose,
    onForward,
    onBackward,
    stepNumber,
    maxStepNumber,
    text,
    boardState,
  } = props;

  // preventOutsideDismiss also disables Escape key, so we enable it manually
  useKeyPressEvent(Keys.Escape, onClose, null);

  return (
    <Dialog
      className={styles.dialog}
      open={true}
      onClose={onClose}
      preventOutsideDismiss={true}
    >
      <DialogTitle>How to play</DialogTitle>
      <DialogContent>
        <p>{text}</p>
        <div className={styles.bigBoard}>
          <HowToPlayContent boardState={boardState} />
        </div>
      </DialogContent>
      <DialogActions>
        <DialogActionBar
          onClose={onClose}
          onForward={onForward}
          onBackward={onBackward}
          stepNumber={stepNumber}
          maxStepNumber={maxStepNumber}
        />
      </DialogActions>
    </Dialog>
  );
}
