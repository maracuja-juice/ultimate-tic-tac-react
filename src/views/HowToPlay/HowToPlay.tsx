import * as React from 'react';
import { useHistory } from 'react-router';
import routes from '../../routes/routes';
import { AppState, HowToPlayBoardState } from '../../state/AppState';
import {
  howToPlayClose,
  howToPlayOpen,
  howToPlayStepBackward,
  howToPlayStepForward,
} from '../../state/howToPlay/howToPlayActions';
import { connect } from 'react-redux';
import { HowToPlayDialog } from '../../components/HowToPlay/HowToPlayDialog';
import { getHowToPlay } from '../../state/selectors/AppStateSelectors';
import { getCurrentHowToPlayText } from "../../state/selectors/howToPlay/HowToPlayStateSelectors";

export interface HowToPlayProps {
  stepNumber: number;
  maxStepNumber: number;
  text: React.ReactNode;
  boardState: HowToPlayBoardState;
  onOpen: () => void;
  onClose: () => void;
  onForward: () => void;
  onBackward: () => void;
}

function HowToPlay(props: HowToPlayProps) {
  let history = useHistory();

  function handleClose() {
    props.onClose();
    history.push(routes.Home);
  }

  return (
    <HowToPlayDialog
      boardState={props.boardState}
      onOpen={props.onOpen}
      onClose={handleClose}
      maxStepNumber={props.maxStepNumber}
      stepNumber={props.stepNumber}
      onBackward={props.onBackward}
      onForward={props.onForward}
      text={props.text}
    />
  );
}

const mapStateToProps = (state: AppState) => ({
  boardState: getHowToPlay(state).boardState,
  stepNumber: getHowToPlay(state).stepNumber,
  maxStepNumber: getHowToPlay(state).maxStepNumber,
  text: getCurrentHowToPlayText(state),
});

const mapDispatchToProps = {
  onOpen: () => howToPlayOpen(),
  onClose: () => howToPlayClose(),
  onForward: () => howToPlayStepForward(),
  onBackward: () => howToPlayStepBackward(),
};

export default connect(mapStateToProps, mapDispatchToProps)(HowToPlay);
