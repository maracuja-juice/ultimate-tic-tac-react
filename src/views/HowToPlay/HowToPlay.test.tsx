import * as React from 'react';
import { shallow } from 'enzyme';
import { HowToPlay } from './HowToPlay';
import { Player } from '../../state/AppState';

describe('HowToPlay', () => {
  it('should match snapshot', () => {
    const howToPlay = shallow(
      <HowToPlay
        onOpen={() => {}}
        onClose={() => {}}
        onForward={() => {}}
        onBackward={() => {}}
        stepNumber={0}
        maxStepNumber={10}
        text={''}
        boardState={{
          board: [],
          activeBoards: [],
          animate: false,
          currentPlayer: Player.Cross,
        }}
      />,
    );
    expect(howToPlay).toMatchSnapshot();
  });
});
