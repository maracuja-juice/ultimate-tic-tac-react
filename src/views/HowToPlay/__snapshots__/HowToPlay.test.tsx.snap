// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HowToPlay should match snapshot 1`] = `
<HowToPlayDialog
  boardState={
    Object {
      "activeBoards": Array [],
      "animate": false,
      "board": Array [],
      "currentPlayer": 0,
    }
  }
  maxStepNumber={10}
  onBackward={[Function]}
  onClose={[Function]}
  onForward={[Function]}
  onOpen={[Function]}
  stepNumber={0}
  text=""
/>
`;
